{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison to saturation genome editing\n",
    "\n",
    "Author: Ruth Hanna\n",
    "\n",
    "This notebook contains the code for Figure 2H, the comparison of data from the BRCA1 base editing screen to the BRCA1 saturation genome editing screen (Findlay et al. 2018). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import base_edit_functions as be\n",
    "from scipy.stats import pearsonr\n",
    "from colors import get_colors\n",
    "\n",
    "mpl.rc('pdf', fonttype=42)\n",
    "mpl.rcParams['font.sans-serif'] = \"Arial\"\n",
    "mpl.rcParams['font.family'] = \"sans-serif\"\n",
    "sns.set_context('paper')\n",
    "sns.set_style('ticks')\n",
    "color_dict = get_colors()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we load filtered LFC files and drop any columns we don't need for this analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(723, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA sequence</th>\n",
       "      <th>sgRNA strand</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Nucleotide edits</th>\n",
       "      <th>Amino acid edits</th>\n",
       "      <th>Mutation bin</th>\n",
       "      <th>AvgHAP1_MELJUSO;LFC_pdna;z-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAACGACCACCCGAAACCT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergenic control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>0.137103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAGGACCCCATATAGCAC</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C_8;</td>\n",
       "      <td>Exon23:+34;</td>\n",
       "      <td>Intron</td>\n",
       "      <td>-0.732224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAGTGGTGGTATACGATA</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>-0.711219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACTTACTTGCAAAATATG</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C_4;C_8;</td>\n",
       "      <td>Exon3:+5;Exon3:+1;</td>\n",
       "      <td>Splice site</td>\n",
       "      <td>-3.033019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAGAGATCATAGAAAAGAC</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>-3.087088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         sgRNA sequence sgRNA strand         Gene symbol Nucleotide edits  \\\n",
       "0  AAAACGACCACCCGAAACCT          NaN  Intergenic control              NaN   \n",
       "1  AAAAGGACCCCATATAGCAC    antisense               BRCA1             C_8;   \n",
       "2  AAAAGTGGTGGTATACGATA    antisense               BRCA1              NaN   \n",
       "3  AAACTTACTTGCAAAATATG    antisense               BRCA1         C_4;C_8;   \n",
       "4  AAAGAGATCATAGAAAAGAC    antisense               BRCA1              NaN   \n",
       "\n",
       "     Amino acid edits Mutation bin  AvgHAP1_MELJUSO;LFC_pdna;z-score  \n",
       "0                 NaN     No edits                          0.137103  \n",
       "1         Exon23:+34;       Intron                         -0.732224  \n",
       "2                 NaN     No edits                         -0.711219  \n",
       "3  Exon3:+5;Exon3:+1;  Splice site                         -3.033019  \n",
       "4                 NaN     No edits                         -3.087088  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load data\n",
    "brca1 = pd.read_csv('../Data/Filtered_LFC_files/Table_S2_filtered_BRCA1_v1.csv')\n",
    "\n",
    "# Drop unnecessary columns\n",
    "brca1 = brca1[['sgRNA sequence','sgRNA strand','Gene symbol','Nucleotide edits',\n",
    "               'Amino acid edits','Mutation bin','AvgHAP1_MELJUSO;LFC_pdna;z-score']]\n",
    "print brca1.shape\n",
    "brca1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we load \"edit-level\" annotations of the library -- this means that each row corresponds to a single predicted edit, and therefore each sgRNA can have multiple rows (if there are multiple predicted edits). We also clean up this dataframe a little, dropping duplicates and unnecessary columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA sequence</th>\n",
       "      <th>Edit nucleotide position(s)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAATCTTAGAGTGTCCCATC</td>\n",
       "      <td>43124034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGCACAAGAGTGTATTAATT</td>\n",
       "      <td>43124005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCACAAGAGTGTATTAATTT</td>\n",
       "      <td>43124005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCAGACTAGCAGGGTAGGGG</td>\n",
       "      <td>43115780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CCCCCCCTACCCTGCTAGTC</td>\n",
       "      <td>43115792</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         sgRNA sequence Edit nucleotide position(s)\n",
       "0  AAATCTTAGAGTGTCCCATC                    43124034\n",
       "2  AGCACAAGAGTGTATTAATT                    43124005\n",
       "3  GCACAAGAGTGTATTAATTT                    43124005\n",
       "4  CCAGACTAGCAGGGTAGGGG                    43115780\n",
       "6  CCCCCCCTACCCTGCTAGTC                    43115792"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edit_annot = pd.read_table('../Data/Annotations/0423_GC_filter_annotations/BRCA1_0422_20-04-22-17-23-50/clinvar_annotations_BRCA1_0422.txt')\n",
    "\n",
    "# Drop duplicates and remove columns related to ClinVar annotations\n",
    "# Dropping duplicates is important, otherwise edits with multiple matches to ClinVar\n",
    "# will have multiple rows in the annotation file\n",
    "edit_annot = edit_annot.drop_duplicates(subset=['sgRNA sequence','Edit nucleotide position(s)'])\n",
    "edit_annot = edit_annot.loc[:,['sgRNA sequence',\n",
    "                               'Edit nucleotide position(s)']]\n",
    "\n",
    "edit_annot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All we really need from this dataframe are the \"edit nucleotide positions,\" so we merge those onto our scores file.\n",
    "\n",
    "One thing we need to be attentive to is that sgRNAs may introduce multiple edits per codon, whereas the SGE screen only assayed single nucleotide variants. So, at least in our first pass of analysis, we don't want to include any sgRNAs that introduce multiple edits per codon, because those weren't assayed in the SGE data.\n",
    "\n",
    "This might be a little overcautious. For example, in many cases an sgRNA may still introduce the same amino acid change, because the second edit in the codon is in the third position and therefore does not affect the coding sequence. However, we'll exclude any of these sgRNAs to be safe, since even edits in the third position could affect splicing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(723, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA sequence</th>\n",
       "      <th>sgRNA strand</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Nucleotide edits</th>\n",
       "      <th>Amino acid edits</th>\n",
       "      <th>Mutation bin</th>\n",
       "      <th>AvgHAP1_MELJUSO;LFC_pdna;z-score</th>\n",
       "      <th>Edit nucleotides</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAACGACCACCCGAAACCT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergenic control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>0.137103</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAGGACCCCATATAGCAC</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C_8;</td>\n",
       "      <td>Exon23:+34;</td>\n",
       "      <td>Intron</td>\n",
       "      <td>-0.732224</td>\n",
       "      <td>43047609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAGTGGTGGTATACGATA</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>-0.711219</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACTTACTTGCAAAATATG</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C_4;C_8;</td>\n",
       "      <td>Exon3:+5;Exon3:+1;</td>\n",
       "      <td>Splice site</td>\n",
       "      <td>-3.033019</td>\n",
       "      <td>43115721_43115725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAGAGATCATAGAAAAGAC</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>-3.087088</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         sgRNA sequence sgRNA strand         Gene symbol Nucleotide edits  \\\n",
       "0  AAAACGACCACCCGAAACCT          NaN  Intergenic control              NaN   \n",
       "1  AAAAGGACCCCATATAGCAC    antisense               BRCA1             C_8;   \n",
       "2  AAAAGTGGTGGTATACGATA    antisense               BRCA1              NaN   \n",
       "3  AAACTTACTTGCAAAATATG    antisense               BRCA1         C_4;C_8;   \n",
       "4  AAAGAGATCATAGAAAAGAC    antisense               BRCA1              NaN   \n",
       "\n",
       "     Amino acid edits Mutation bin  AvgHAP1_MELJUSO;LFC_pdna;z-score  \\\n",
       "0                 NaN     No edits                          0.137103   \n",
       "1         Exon23:+34;       Intron                         -0.732224   \n",
       "2                 NaN     No edits                         -0.711219   \n",
       "3  Exon3:+5;Exon3:+1;  Splice site                         -3.033019   \n",
       "4                 NaN     No edits                         -3.087088   \n",
       "\n",
       "    Edit nucleotides  \n",
       "0                     \n",
       "1           43047609  \n",
       "2                     \n",
       "3  43115721_43115725  \n",
       "4                     "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_edit_nucs(sgrna):\n",
    "    edit_nuc = edit_annot.loc[edit_annot['sgRNA sequence'] == sgrna,'Edit nucleotide position(s)'].tolist()\n",
    "    # We join this list with \"_\" so that we can easily identify any sgRNAs with multiple edits/codon\n",
    "    # (They will contain \";\")\n",
    "    return '_'.join(edit_nuc)\n",
    "    \n",
    "brca1['Edit nucleotides'] = brca1['sgRNA sequence'].apply(get_edit_nucs)\n",
    "print brca1.shape\n",
    "brca1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we remove any sgRNAs that contain multiple edits per codon (as evidenced by the presence of \";\" in the \"Edit nucleotides\" column). We also remove any sgRNAs predicted to make no edits, since those won't be helpful in our comparison to the SGE dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(272, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA sequence</th>\n",
       "      <th>sgRNA strand</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Nucleotide edits</th>\n",
       "      <th>Amino acid edits</th>\n",
       "      <th>Mutation bin</th>\n",
       "      <th>AvgHAP1_MELJUSO;LFC_pdna;z-score</th>\n",
       "      <th>Edit nucleotides</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAGGACCCCATATAGCAC</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C_8;</td>\n",
       "      <td>Exon23:+34;</td>\n",
       "      <td>Intron</td>\n",
       "      <td>-0.732224</td>\n",
       "      <td>43047609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACTTACTTGCAAAATATG</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C_4;C_8;</td>\n",
       "      <td>Exon3:+5;Exon3:+1;</td>\n",
       "      <td>Splice site</td>\n",
       "      <td>-3.033019</td>\n",
       "      <td>43115721_43115725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AAATAACAAGTGTTGGAAGC</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C_7;</td>\n",
       "      <td>Leu1229Leu;</td>\n",
       "      <td>Silent</td>\n",
       "      <td>-0.083921</td>\n",
       "      <td>43091844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AAATCTTAGAGTGTCCCATC</td>\n",
       "      <td>sense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C_5;</td>\n",
       "      <td>Ile21Ile;</td>\n",
       "      <td>Silent</td>\n",
       "      <td>0.146696</td>\n",
       "      <td>43124034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AACCAATCAAGAAAGGATCC</td>\n",
       "      <td>antisense</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C_4;C_8;</td>\n",
       "      <td>Gly1319Ser;Leu1317Leu;</td>\n",
       "      <td>Missense</td>\n",
       "      <td>-0.858230</td>\n",
       "      <td>43091576_43091580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sgRNA sequence sgRNA strand Gene symbol Nucleotide edits  \\\n",
       "1   AAAAGGACCCCATATAGCAC    antisense       BRCA1             C_8;   \n",
       "3   AAACTTACTTGCAAAATATG    antisense       BRCA1         C_4;C_8;   \n",
       "9   AAATAACAAGTGTTGGAAGC    antisense       BRCA1             C_7;   \n",
       "16  AAATCTTAGAGTGTCCCATC        sense       BRCA1             C_5;   \n",
       "19  AACCAATCAAGAAAGGATCC    antisense       BRCA1         C_4;C_8;   \n",
       "\n",
       "          Amino acid edits Mutation bin  AvgHAP1_MELJUSO;LFC_pdna;z-score  \\\n",
       "1              Exon23:+34;       Intron                         -0.732224   \n",
       "3       Exon3:+5;Exon3:+1;  Splice site                         -3.033019   \n",
       "9              Leu1229Leu;       Silent                         -0.083921   \n",
       "16               Ile21Ile;       Silent                          0.146696   \n",
       "19  Gly1319Ser;Leu1317Leu;     Missense                         -0.858230   \n",
       "\n",
       "     Edit nucleotides  \n",
       "1            43047609  \n",
       "3   43115721_43115725  \n",
       "9            43091844  \n",
       "16           43124034  \n",
       "19  43091576_43091580  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered = brca1.loc[~brca1['Edit nucleotides'].str.contains(';')]\n",
    "filtered = filtered.loc[filtered['Mutation bin'] != 'No edits']\n",
    "print filtered.shape\n",
    "filtered.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we're ready to compare our data to the SGE data, obtained from the original publication. We've already converted the genomic positions to GRCh38 using liftOver."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3893, 65)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position (hg19)</th>\n",
       "      <th>position (hg38)</th>\n",
       "      <th>reference</th>\n",
       "      <th>alt</th>\n",
       "      <th>snp_change</th>\n",
       "      <th>snp</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>...</th>\n",
       "      <th>aGVGD.diff</th>\n",
       "      <th>aGVGD.class</th>\n",
       "      <th>clinvar</th>\n",
       "      <th>clinvar_simple</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "      <th>bravo_AF</th>\n",
       "      <th>flossies_AF</th>\n",
       "      <th>WT.HAP1.function.score.r1</th>\n",
       "      <th>WT.HAP1.function.score.r2</th>\n",
       "      <th>WT.HAP1.function.score.mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17</td>\n",
       "      <td>41276135</td>\n",
       "      <td>43124118</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>43124118:T&gt;G</td>\n",
       "      <td>NM_007294.3</td>\n",
       "      <td>-22</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.369446</td>\n",
       "      <td>0.259903</td>\n",
       "      <td>0.314675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17</td>\n",
       "      <td>41276135</td>\n",
       "      <td>43124118</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>43124118:T&gt;C</td>\n",
       "      <td>NM_007294.3</td>\n",
       "      <td>-22</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Conflicting interpretations of pathogenicity</td>\n",
       "      <td>Conflicting interpretations of pathogenicity</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.338637</td>\n",
       "      <td>0.301356</td>\n",
       "      <td>-0.018641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17</td>\n",
       "      <td>41276135</td>\n",
       "      <td>43124118</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T&gt;A</td>\n",
       "      <td>43124118:T&gt;A</td>\n",
       "      <td>NM_007294.3</td>\n",
       "      <td>-22</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.975176</td>\n",
       "      <td>-0.101218</td>\n",
       "      <td>-0.538197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17</td>\n",
       "      <td>41276134</td>\n",
       "      <td>43124117</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>43124117:T&gt;G</td>\n",
       "      <td>NM_007294.3</td>\n",
       "      <td>-21</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.766777</td>\n",
       "      <td>0.529892</td>\n",
       "      <td>-0.618443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17</td>\n",
       "      <td>41276134</td>\n",
       "      <td>43124117</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>43124117:T&gt;C</td>\n",
       "      <td>NM_007294.3</td>\n",
       "      <td>-21</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene  chromosome  position (hg19)  position (hg38) reference alt  \\\n",
       "0  BRCA1          17         41276135         43124118         T   G   \n",
       "1  BRCA1          17         41276135         43124118         T   C   \n",
       "2  BRCA1          17         41276135         43124118         T   A   \n",
       "3  BRCA1          17         41276134         43124117         T   G   \n",
       "4  BRCA1          17         41276134         43124117         T   C   \n",
       "\n",
       "  snp_change           snp transcript_ID transcript_position  \\\n",
       "0        T>G  43124118:T>G   NM_007294.3                 -22   \n",
       "1        T>C  43124118:T>C   NM_007294.3                 -22   \n",
       "2        T>A  43124118:T>A   NM_007294.3                 -22   \n",
       "3        T>G  43124117:T>G   NM_007294.3                 -21   \n",
       "4        T>C  43124117:T>C   NM_007294.3                 -21   \n",
       "\n",
       "              ...             aGVGD.diff aGVGD.class  \\\n",
       "0             ...                    NaN         NaN   \n",
       "1             ...                    NaN         NaN   \n",
       "2             ...                    NaN         NaN   \n",
       "3             ...                    NaN         NaN   \n",
       "4             ...                    NaN         NaN   \n",
       "\n",
       "                                        clinvar  \\\n",
       "0                                        absent   \n",
       "1  Conflicting interpretations of pathogenicity   \n",
       "2                                        absent   \n",
       "3                                        absent   \n",
       "4                                    Pathogenic   \n",
       "\n",
       "                                 clinvar_simple gnomAD_AF  bravo_AF  \\\n",
       "0                                        absent  0.000000  0.000000   \n",
       "1  Conflicting interpretations of pathogenicity  0.000127  0.000167   \n",
       "2                                        absent  0.000000  0.000000   \n",
       "3                                        absent  0.000000  0.000000   \n",
       "4                                    Pathogenic  0.000000  0.000000   \n",
       "\n",
       "  flossies_AF WT.HAP1.function.score.r1  WT.HAP1.function.score.r2  \\\n",
       "0         0.0                  0.369446                   0.259903   \n",
       "1         0.0                 -0.338637                   0.301356   \n",
       "2         0.0                 -0.975176                  -0.101218   \n",
       "3         0.0                 -1.766777                   0.529892   \n",
       "4         0.0                       NaN                        NaN   \n",
       "\n",
       "  WT.HAP1.function.score.mean  \n",
       "0                    0.314675  \n",
       "1                   -0.018641  \n",
       "2                   -0.538197  \n",
       "3                   -0.618443  \n",
       "4                         NaN  \n",
       "\n",
       "[5 rows x 65 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sge_data = pd.read_table('../Data/External_datasets/BRCA1_BRCA2/SGE_BRCA1.txt')\n",
    "print sge_data.shape\n",
    "sge_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We drop some unnecessary columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position (hg38)</th>\n",
       "      <th>reference</th>\n",
       "      <th>alt</th>\n",
       "      <th>protein_variant</th>\n",
       "      <th>function.score.mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>43124118</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.372611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>43124118</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.045313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>43124118</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.108254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>43124117</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.277963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>43124117</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.388414</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position (hg38) reference alt protein_variant  function.score.mean\n",
       "0         43124118         T   G             NaN            -0.372611\n",
       "1         43124118         T   C             NaN            -0.045313\n",
       "2         43124118         T   A             NaN            -0.108254\n",
       "3         43124117         T   G             NaN            -0.277963\n",
       "4         43124117         T   C             NaN            -0.388414"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sge_data = sge_data[['position (hg38)','reference','alt','protein_variant','function.score.mean']]\n",
    "sge_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, for each row in the sgRNA dataframe, we get the function scores corresponding to each of the predicted edits. For example, if an sgRNA is predicted to make 3 edits, we would retrieve 3 SGE scores (one for each edit). We then return the minimum function score, reasoning that the edit with the strongest effect should drive the phenotype of the sgRNA.\n",
    "\n",
    "It's worth noting that this is an imperfect assumption (for example, two edits may interact), but it is a reasonable way to get to a single score per sgRNA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of sgRNAs with combined SGE score = 51\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Amino acid edits</th>\n",
       "      <th>AvgHAP1_MELJUSO;LFC_pdna;z-score</th>\n",
       "      <th>Edit nucleotides</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Mutation bin</th>\n",
       "      <th>Nucleotide edits</th>\n",
       "      <th>min_sge_score</th>\n",
       "      <th>sgRNA sequence</th>\n",
       "      <th>sgRNA strand</th>\n",
       "      <th>sge_variants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Exon3:+5;Exon3:+1;</td>\n",
       "      <td>-3.033019</td>\n",
       "      <td>43115721_43115725</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Splice site</td>\n",
       "      <td>C_4;C_8;</td>\n",
       "      <td>-2.398164</td>\n",
       "      <td>AAACTTACTTGCAAAATATG</td>\n",
       "      <td>antisense</td>\n",
       "      <td>nan;nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ile21Ile;</td>\n",
       "      <td>0.146696</td>\n",
       "      <td>43124034</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Silent</td>\n",
       "      <td>C_5;</td>\n",
       "      <td>-1.601567</td>\n",
       "      <td>AAATCTTAGAGTGTCCCATC</td>\n",
       "      <td>sense</td>\n",
       "      <td>p.I21=</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Glu29Lys;Leu28Leu;</td>\n",
       "      <td>-0.651837</td>\n",
       "      <td>43115776_43115775</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Missense</td>\n",
       "      <td>C_5;C_6;</td>\n",
       "      <td>-0.645512</td>\n",
       "      <td>AACTCCAGACTAGCAGGGTA</td>\n",
       "      <td>antisense</td>\n",
       "      <td>p.L28=;p.E29K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Thr1661Ile;</td>\n",
       "      <td>-3.912033</td>\n",
       "      <td>43070995</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Missense</td>\n",
       "      <td>C_7;</td>\n",
       "      <td>-0.021899</td>\n",
       "      <td>AATTGACAGCTTCAACAGAA</td>\n",
       "      <td>sense</td>\n",
       "      <td>p.T1640I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Exon22:+10;</td>\n",
       "      <td>-1.092252</td>\n",
       "      <td>43049111</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Intron</td>\n",
       "      <td>C_4;</td>\n",
       "      <td>-0.376090</td>\n",
       "      <td>ACCCAATACTTACTGTGCCA</td>\n",
       "      <td>antisense</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Amino acid edits  AvgHAP1_MELJUSO;LFC_pdna;z-score   Edit nucleotides  \\\n",
       "3   Exon3:+5;Exon3:+1;                         -3.033019  43115721_43115725   \n",
       "16           Ile21Ile;                          0.146696           43124034   \n",
       "24  Glu29Lys;Leu28Leu;                         -0.651837  43115776_43115775   \n",
       "59         Thr1661Ile;                         -3.912033           43070995   \n",
       "75         Exon22:+10;                         -1.092252           43049111   \n",
       "\n",
       "   Gene symbol Mutation bin Nucleotide edits  min_sge_score  \\\n",
       "3        BRCA1  Splice site         C_4;C_8;      -2.398164   \n",
       "16       BRCA1       Silent             C_5;      -1.601567   \n",
       "24       BRCA1     Missense         C_5;C_6;      -0.645512   \n",
       "59       BRCA1     Missense             C_7;      -0.021899   \n",
       "75       BRCA1       Intron             C_4;      -0.376090   \n",
       "\n",
       "          sgRNA sequence sgRNA strand   sge_variants  \n",
       "3   AAACTTACTTGCAAAATATG    antisense        nan;nan  \n",
       "16  AAATCTTAGAGTGTCCCATC        sense         p.I21=  \n",
       "24  AACTCCAGACTAGCAGGGTA    antisense  p.L28=;p.E29K  \n",
       "59  AATTGACAGCTTCAACAGAA        sense       p.T1640I  \n",
       "75  ACCCAATACTTACTGTGCCA    antisense            nan  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_sge_scores_v2(row):\n",
    "    edit_nucs = row['Edit nucleotides'].split('_')\n",
    "    scores = []\n",
    "    variants = []\n",
    "    not_found = False\n",
    "    for nuc in edit_nucs:\n",
    "        df = sge_data.loc[sge_data['position (hg38)'] == int(nuc),:]\n",
    "        # If ANY of the edit nucleotides are not found in SGE data, return np.nan\n",
    "        if len(df) == 0:\n",
    "            not_found = True\n",
    "            row['min_sge_score'] = np.nan \n",
    "            row['sge_variants'] = np.nan\n",
    "        # Else, get mean function score for the C>T or G>A edit and append protein variant\n",
    "        if list(set(df['reference'].tolist())) == ['C']:\n",
    "            # Get function score and variants\n",
    "            found_scores = df.loc[df['alt'] == 'T','function.score.mean'].tolist()\n",
    "            if len(found_scores) == 0:\n",
    "                not_found = True\n",
    "            else:\n",
    "                scores.extend(found_scores)\n",
    "                variants.extend(df.loc[df['alt'] == 'T','protein_variant'].tolist())\n",
    "        elif list(set(df['reference'].tolist())) == ['G']:\n",
    "            # Get function score and variants\n",
    "            found_scores = df.loc[df['alt'] == 'A','function.score.mean'].tolist()\n",
    "            if len(found_scores) == 0:\n",
    "                not_found = True\n",
    "            else:\n",
    "                scores.extend(found_scores)\n",
    "                variants.extend(df.loc[df['alt'] == 'A','protein_variant'].tolist()) \n",
    "    # Return the minimum SGE score\n",
    "    if not not_found:\n",
    "        # Handle cases where scores is empty (e.g. nucleotide was assayed but C>T or G>A variant was not)\n",
    "        if len(scores) == 0:\n",
    "            row['min_sge_score'] = np.nan\n",
    "        else:\n",
    "            row['min_sge_score'] = min(scores)\n",
    "        # Handle np.nan for non-coding variants    \n",
    "        variants = [str(variant) for variant in variants] \n",
    "        row['sge_variants'] = ';'.join(variants)\n",
    "    return row\n",
    "\n",
    "filtered = filtered.apply(get_sge_scores_v2,axis=1)\n",
    "# Drop rows that don't have SGE score\n",
    "filtered = filtered.dropna(subset=['min_sge_score'])\n",
    "print 'Total number of sgRNAs with combined SGE score = ' + str(len(filtered))\n",
    "filtered.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we correlate the sgRNA score with the minimum SGE score and observe a modest correlation (Pearson r = 0.44). We draw lines at the thresholds for LOF, intermediate, and functional SGE scores, as described in the original publication. This comparison is shown in __Figure 3H__. Colors indicate the predicted mutation bin of the sgRNA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pearson r = 0.44\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAKEAAACWCAYAAAC2LuijAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAHRhJREFUeJztnXl4VOXZ/z+zb5lMyAIJS8KSgFA2IWqjUhSU5a1aLT83+NHSgljFVq5L+6pFJYra+NalItVeUSq98CcI0mLhFUXQWGQtYQsCJrKE7BtZJpPMes7vj+hAIMksyZyZSc7nuriuM3PO/Tz3mXx5zjn3uZ/nVoiiKCIjE0aU4XZARkYWoUzYkUUoE3ZkEcqEHVmEMmFHFqFM2JFFKBN2ZBHKhB1ZhDJhRxahTNhRh9uBrli2bBnJycnhdiMsVFVVMWDAgJDbhIrKykpefPFF/w4WI5iVK1eG24WwsXz5cklsQkUgfzv5chyh3HTTTZLYXI6n2YnjXAOC09Pttvwloi/HfZny8nJJbC7FY3VQ+/fDiK1u1MkxJPzfCSgUim616Q/ySBihFBYWSmJzKa7KZsRWNwDuS7ZDTa8XoVuw81VpDv86/TBnG78Ktzt+s3jxYklsLkU7OBZVPwMAuvR4FAZpLpS9XoTnmnZRbsvH5q7hP1XvhNsdv8nNzfXrOI/VgbvR7rVx1dho2nGa1uNVAfepNGhIXHA1iYszibtztCSXYugD94QmTdIl24lh9CQwUlJSfB5j/66Oho9PgSASe+sIUpJTqN94HMHmAkBp1qFLiwNAFEQUyvaiEt0CqBTtxKZQK1Fb9D14Jr7p9SJMMU3kxoGP0eAoZrhlerjd8ZvMzEyfx9hP1YLQNjuj9VQNmZMnI+xyePcLLW1ibDlWSdPnp1GatMTfOxZ1PwO2A6VYvzqHyqIj/r5xqGKlFd6l9PrLMcAQ848Zl3hvVI2EW7Zs6XK/s6kST8p5xLgmAPTpCWz5361YZo9EnWTCMG4A+lGJiKJAU9EBhNRSPPZmWo9XA2DdexYAT6OD1hPVoT0ZH/T6kTBamTFjRqf7BJedum8+RhTcMEZB3NC7MAxKYYZzBobRSRhGX7wFaS4/hpBcAoBotKNJHo/HYUM0WlE4YxEREc2tIT+frugTI2GkUlvfypFT1TRYHVfs6yrcIgiuNgECIKIyKzq1EVx277YqSYM+IwGlRo9ifC3CyGKYcAbDCN/3n6FEHgnDhNXmZN0nJ3G5BYx6Nb+6axw6rcq7/9y5c53aqnVmzGlZtFafQtcvFW1scqc2MQMn4G6pw+O0YRk2BQCFUkXSxJ9jry9Ga05GrY/t0XMLFElFWFVVRU5ODhaLhYyMDObNmydl9xFFU7MDl1sAoMXupsXuaidCXzE/8+BJmAdPAsBeWIuzzMriexZccZxSoyd+9H9d8b1KF4Mp+Uc0HipHvHCamMkDUX8fI5QaSS/H69evZ/78+WRnZ5OXl4fL5ZKy+4giJSmG9NQ4lEoFYzMS6XfZ06m/cUJnWRMNH5+i5WAZ9q2nEQX/1zIo/OosrTvPYD9cQc2G4wH535NIOhLW1tZ641+xsbFYrVbi4+O9+/fv38+BAwe8n0tLS6V0T1KUSgV33Jze6f6hQ4f61Y5gc3q39QoteARQqq44rqTSyr//fQaTzcXUtHj6TRtO2fl6Yi9pR3A78TitqA3xkgWqQWIRpqSkUFlZSUpKCo2NjcTGtr8Xue6667juuuu8n998800p3YsoRo4c2eX+rw+VcurMBdJT47h63ACcZU3U9LMzUHOlAAE+/fcZrK0uUELciSpuGJ2EelQi5+pLsXhEtJMSqT78AYLThj5+WIeX8FAh6eX47rvv5v333+fZZ59lxowZqNXyc1FnbN++vdN9FxpbOVBQSZPNyaGT1divGUjSwslsOvZ5pzbaS8SpARQGDVOuTSXl/4yl3/yJpIwGwWkDwH7hLIJHulslv1QgCAJKZff1mpSUxKuvvtrtdvoCt99+e6f79Fo1arUSt1tApVJg1Gt82twxPZ39+0swtTiZ/JPhqOP0OEsbSaq3ox+ViCCkoFTrEdx2dHFDUKo0fvlptTlpsbsYkGAK7AQvwacIc3NzOXLkCDfffDNWq5Vf//rXQXcm4z8HDx5k8uTJHe4zGjTcPWMkp0saGDbIgtmkBSD/P/kMP2/AWdKIafIgYm5I9dr0i9Uz69YM7+fWMxdo2HQCBVBxoIQRCzNJmjQXj70RjSnp8i47pKzaykefFeIRRK4e3Z+br031bdQBPoe3iooK0tPTufvuuykqKgqqk76Oo6mC2uMf03D6K0TBv4zlioqKLvenJMVw46TBDBpg9n6nqXHi+O4CosND857zCPbO8wHPHavkh0cPVb0dq82JSmNAa05G0cGDTYdtlDXh+f5p/LvzDX7ZdIRPEYqiiMfjobS0FKvVGnRHfZmGws9xNpbSUnmclqoTftkEkhvoOFdP3ftHmZV+PT8oS2nWotBc/PO2tLrYd7ScU2cvAHDU7aZRpcADfGPWoKixIXoEv/sEGDEkDo26rY/Rw+N9HN05Pi/H9913H7m5ubzyyis8/PDDQXfUl7l0ZPF3lMnNzSU7O9uvYxs/KUSwuVAAxsyBqExadKMSUaguivDjL7+joqbtwUOpgLhkM/9rc4IoMt7uoXnjN7iG96PfnB/5fV7JiSYWzhmH3eEhvhvpXz5F+MEHH/Daa68F3YEM9LtqFs2lh1Eb4jD0H+2Xja8QzaUodGr4PodQMyAGw5j+7faLgoeGBis/XPgamx3ckpXGoAExKCuaSdhfBoDjTD2iKAYUIzTqNd4Ho2DxKcLjx4/z2GOPYTa33Xv4+79T5iIaYwL9Rt7S6X5RFEGkXdLpwIED/W6/311jsB0so7iuDM2xShq3FWGcPJDYm4YB0FL9LZMSTnO4NhWz1sm4jCTUKiXjMpLwDDBTd7wGweZEPyZJ0iD1D/gU4ZtvvonNZkMURWJiYqTwKeI4WvMB1a0nyIibydDYKT3atquqmfqPvkFweoi7bRT6jAQA8vLy/J7CqY43YJmRzqEVn9FfPwGAlv+UEfPjISj1apQaPUNiLjAk5gIqfSx63cU/uypWT+KiyQg2Jx6bkwsfFqCKMxA7fTgKtTRhZJ8i/Oyzzzh06BBKpZLMzEx+8YtfSOFXxFBhO8qJC/8EoK71OwaZJqNRGdsf1FoATf8Cw9UQG9ibhpbDFd4MaNuBUq8I58yZE7Cv195yA+y1gUdEZdGh+D4hwpAwHGHETbhb6jAlj7vCTqlVodQaqN94HE+jA843ok4yYprk/2jcHXyKsLi4mFWrVgGQk5MTcociDY3y4g23SqFBobjswUJwQtnDIFihcSNoBoFhgv/tD4ihtaBtUpIm+eKVJi8vj3HjrhRMV+w8uIvfzPsVrgpr22y5Sy7vpuS2Bw7B7aC1tgiNKQm1Ia59A5eMfFKNguCHCK1Wqzc0U1dXF3KHIo1EwyiuS36Y6pZvGG6Zhlqpu+wINwgtFz96mgJq33h1CiqLDsHpQT/y4vSDYH7ruro6NANi0Ay48rZJFEXsp2poOr8HT0wZCqWapIn3thNiv5+Npnl/Kep+egzjpFvTxqcIFy9ezPLlyxFFsdvzWqOV4ZabGW65ueOdSiMMWA4N69oux6YbA25f10GMrafnHTf/uxjbgVIgEVJdiIOrcdlq24lQnWAk7r/8fyrvKXyOuUVFRSxatIgZM2Zw5swZKXyKXEQXVOdA6W+g9QgAda1FlKpSEFLfh/7/DT30dOlvPqG/Nq7qZu+2wmZAbUpAFzckKN96Gp8i3L17N2PGjGH27Nns2rVLCp8il8bN0LAeWvZBxVOUWg+w/fwydpX9D/+pClw0XRHo/eDlNm57E3XfbKHumy247U2Yrh2MQqdCYVATf+v1JE24F6X68luL8ODzcqxQKLwhGrGvVyBTaNpt17SeAtp+k7btniOYcNilNtbifTgazgNQf2gnidfcQf/f/jgscUBfqLJ9RJ9TU1N59dVX2bFjBwsWLJB0EcYNGzawd+9eEhIS2LBhA9u2bWPkyJG8+uqr1NXVUVJSwvvvv09KSgp///vfycvLY8iQIfz5z3/GarVSWFjIunXrGDp0KG+99RYHDhwgISGBVatW4XA4OHr0KBs2bPC2WVBQgNFo5O233wZgz549bNq0ybv/22IYOHAQZ0ttVHh+wbGjtYhxZSjVcH5fDN8eqqKlpYXVq1djNpvZsmULW7du9dqXl5dTV1fHmjVrfJ7TkSNHOHnyZEDnlJeXhyiKbNiwgQkjU8DZllQgVGnI/9dJioXaK8/p229RKBTk5uai1WrZuXMnmzdv9u4/d+5cUOdkNBrbJSh3ia8FDI8cOSI2NDSIOTk54s6dO4NZLzFoomGRTI/gEp2elh5v99SpU92y8bgcYtXOLWLZhx+K5a/miXUfFvSkez7p0UUy165dy6FDh0hKSuLLL78MfljrpSgVajTKnp+ltnXr1m7ZKNVa4idOR9uagdpsQnR5qP/4JJ5L5qRECj5FaDKZ2L17N9OnT8dms0nhkwzgcFw5IT5QG3W8gcRfXo063oCr3IqjsI7m3ed7ysUew+eDyY033khlZSU2m41bb71VCp9kgLlz5/acTZjehPiLT49uvfVW5s+fz/bt25k9e7YUPskAa9as6TGb2FtGYJiQjHHyQGJuTOueYyHA7+lu589H3jDem+lsfkkwNiqTFsuMzuc4hxufI+EPqyTcf//9VFeHdwkxmd6JTxHOnDmTwsJCrrnmGlavXi2FTxGP02Mjv+pvHKx6F4cnNPNu8vPzJbGJBHxejqdNm8Y777zDz3/+84iMtoeDozX/j+8a2yaau4RWslJ+2+N9LFiwQBKbSMDnSGgwGMjJyWHbtm3s379fCp8iHrd4MdbmEQIPpfjDBx98IIlNJOBzJLzttttQqVQ8//zzbNiwQQqfOsXdUo9CrUWlDX62f08wIWkubqEFURS4uv8vQ9KHThd4ckEwNj8gONrmKCt10i/N4rPHUaNGebfvueeekDrTFdbz/8FacgCUKhLG3IHOIk3qeUcY1fFMGfTfIe3jtttuk8QGwHG2nvrNJ0EQ2+a5jJJ2be/Ii1x2Qmvt96s/CB7sF3p/XuO6desksYG21f1xCyCIbdsSE5AIv/76a44dOxYqX7pEnzC8bUOhRN9vaFh8kJKsrCxJbAB0qRezq7WpcV0cGRoCugGw2+0kJCSEypcuiU37MYaEESg0etQ6s2+DKKe5udn3QT1gA23zXNT9TSCKaAdbgmqjO/gcCT/55BPvP5VKRVpa+F77aGKS+oQAAQoKCiSxgbZJUEKzE4/VGVTishDAEsUd4XMk3LhxI1lZWSiVSr744gvy8vJ47rnnutWpjG+kLLBo21dK89fFALirbZinDvXLrrnFycbPvqXB6mBq5hAmjQku4dnnSJiens7ixYtZtGgREydO7LOrMEhNT0906gp3zcUUPVeN/+l6p85coL7JgSjC3qPB11r2md6/Y8cOSktLKSgooKSkBKVSyU9+8pOgOwyEiEvvD1EqfEfn5Ha7OXz4cEDnVF1djd1uD/icYlMSUJS3oNSo2e85zbotHzFyaDI7/rmaE6e+pcFq7/CcNmz8CK15MKDA0VxNrM4emvR+j8cj7tmzR8zLyxM9Ho/Y0NDQjaTvwIiG9P5QcezYMUlsOsLjbBXL9+aKZV+vEsv3/FV025s7PbasyioeL6oRnU53u+97NL3/jTfeIDc3l7Vr1/Lggw9isUj/9NQX2bRpkyQ2HeFxtSB62l5NioIbj7Pzp+6B/WP4UXoimk6qBviDzwcTURR57733gu5AJjj8XZGruzYdIVQpUH0zHkHdjO7HWjQx/X0bdQOfImxubmbfvn0YjW0rUY0fPz6kDsm0UV4e+I1+MDYd0fhZEaJVgQIzOtvQkGdP+RRhfHw8Bw8e9H6WRSgNXVX57EmbjlAaNQhW5/fb2h5psyu6FOGWLVvQai86IecTSoeUccLL6XfnaGz5Fajj9OjH+FdOojt0+WCSnp7OmDFjmDhxIhMnTmT48OEhd0imDSnjhJejitUTe/MwjFenSDLwdDoS1tXVcfbsWT799FNmzZoFtD19TZ8+PejOVq1aRWFhIUajkVmzZvXYjXRv5IdClKG2iQQ6FaHZbMbpdKLT6XA62+4PHnrooW51dvjwYVJTU3E6nfzoR1eWKuhLVT59kZmZKYlNJNCpCLVaLSqVioyMDO8su5qamoBO9MMPP2y3NMX06dP55S9/SVFREStXrmTFihXtju+qyqe7tYGm4v2otAZi065HoerdxRm3bNkS8LTPYGwigS7/khkZGSQnJwfd+L333su9994LtMUb33nnHRQKBXFxcQEX3K4v2onL2pZwqdQYMQ+Jzv/1/jJjxgxJbCKBLkX4yiuvoFAoUKnaouEqlYprrrkmqI4UCgUej4fly5fT2toaxKVd7GS7d1JYWMj1118fcptIoEsRvvvuu+Tk5PC73/0OpVLJ66+/3q3OunNPGZc+HWvxPpRaI6aBE7vlRzRw7tw5SWwiAZ83Vo2NjRiNRkRRDDpztyfQGPsRP7rvrIUTzjih1PhMYJg9ezYPPfQQjzzySNCzuXorzc4qtp75HRuLfsF5694ebTuccUKp8TkSDhs2jGnTpjFy5MiojUOFiu8aP8fqaqtLXFD7Ianm4CYadcTQoUMlsYkEfI6Eb775JidPnkSv18tp/Zdh0V4swWDRBVf1vDMCqfLZHZtIwOdIaLFYMBgMjBo1isREaSdFRzrDLFPRqky0uusZGtuz2ebbt28P+Ek3GJtIwKcI4+Li2LVrF08++SSxsbFS+BRVDIoJTbzy9ttvl8QmEvApwiVLljBr1ixEUSQ9PXIXWuxtHDx4MOC3H8HYRAJ+vfsaMWJEqP2QuYyKigpJbCKBqFmLJppwX2jFursYx9n6oNuQ44SXsHHjRn7/+9+zadOmoGpr9DVEj8CF9cew7SmhftM3uCqDC/D3pTihTxEeP36clJQU7rrrLrmYjh+IHgHB9n1yhgieJntQ7fSlEI1PEf6QvOBwOALOfOmLKLVqzDcNQxmjRX9VIroRV9Yy9oeBAwNffzEYm0jArzomx44dY/Hixd60LJmuMV0ziP4PXUvc7VehUAV3252XlyeJTSTg8xcaPHgwb7/9NrNmzcJs7hsrYkUCc+bMkcQmEvApwtdff538/HzKysqidmHuaEQeCS8hMTGR/Px85syZI0/5lJC6ujpJbCIBnyJMS0vD4XBgs9m46qqrpPBJBjlO2I7bbruNadOm4XA4ovbpKxrpS3FCn6/tHn/8cRQKBWlpaZw+fVouNysR48aNk8QmEvApwtTUVKZMmUJsbCyff/65FD5FFR7BiUd0oVX1bIGfYFbEjdZVdH1ejltaWjAajaxdu5aioiIpfIoa6u3n2Hz6QTZ99ysK67f1aNt79wY+XSAYm0jA50j43HPPIQgCer0ekym85bwijXNN/8YptL0bLqzfxsh+PTcR6/7775fEJhLwORKuXbuWpUuX8u6777Jv3z4pfIoakoyjgbawVX/jlcuadIdgkkWiNcHE50hYUVHBX//6VwDeeuutkDsUTQyOuYaZaTnY3Q2kmHp2LrTDEXj10GBsIoEuRXj06FFqamrYu3cvBoOBsrIyqfyKGuL1oVkub+7cuZLYRAJdXo537dpFRkYG+fn5fP311/KUTwlZs2aNJDaRQJcj4SOPPCKVHzKXEcxckWicXwJyer9MBCCLMELJz8+XxCYSkEUYoSxYsEASm0hAFmGEEkzuZrTme8oijFB0Op0kNpGALMIIJZhl+KJ16b6oEaHbYaXhuy9pKt6PKHjC7U7IWbdunSQ2kUDULIHfULgDZ1Nb7TaFSoN58KQwexRasrICX+swGJtIIGpGQtHj6nC7txLM0szhXM65O0SNCOPSb0ZrGYQhMYOYQb1/4fSCggJJbCKBqLkca2KSSBx7Z7jdkAx5opNM2OlLE51kEUYoCQkJkthEAiG9HP/zn/+koKCAZ599FpvNxvLly4mLi8NsNvPoo4+GsuuoJ5gKqNFaNTVkI+Enn3xCTU0NbrcbgG3btjF16lSefvppKioqqKqqClXXvYJNmzZJYhMJ9NhIeHlFzz/84Q+MHz/ee59SW1vrrRA6YMAAqqurGTBgQLs2Li81W1BQ0K7SJ7SVnx08eHBPuR1ygvU3ISHhinMPhc3l9NTvW1lZ6f/BYggpKSkRn3nmGVEURXHz5s3ili1bRFEUxSeeeEKsrKwMqs2VK1f2mH9SIPvrG8lCNDNnzuSZZ57h0KFDpKamXjEKyvRdQirCwYMH8/zzzwOg1+v505/+FMruZKKUqAvRXHvtteF2ISBkf32jEEWx91ewloloom4klOl9yCKUCTtRk8BwKRUVFcybN48vvvgi3K74ZNWqVVRWVtLQ0MADDzzAhAkTwu1Sh1RVVZGTk4PFYiEjI4N58+ZJ1nfUjYStra28/fbbDBo0KNyu+EQURYYNG8YLL7zAb37zG3bs2BFulzpl/fr1zJ8/n+zsbPLy8iStWRPxI+Glb2JEUSQmJobs7GyWLVsWZs86pqM3R7W1tbz77rs89dRTYfSsa2pra73LvMTGxmK1WomPD64QUMBIHh7vBidPnhQffPBB8ZlnnhGnTp0q/u1vfwu3Sz45ePCg+OSTT4p1dXXhdqVL/vKXv4iHDh0SRVEUFy5cKLpcLsn6jtoQzcKFC1m9enW43egSq9XKzJkzycrKQqFQcN1113H33XeH260OqampIScnB5PJxNixY7nnnnsk6ztqRSjTe4i6BxOZ3ocsQpmwI4tQJuzIIpQJOxEfJ+wJLly4wB//+Ed0Oh02m4158+aRmZnJnj172LhxIwaDAbvd7i2hdscdd7Sror5kyRKGDRvWbT/+8Y9/oNPp+OlPf+r97uWXX+aJJ54IqJ1oiAwEQp8Q4cmTJxk0aBCPPvooLpeLnTt3YrVaWb16Nbm5ud7q9j+80UhKSuKVV17psK1vv/2WNWvW4PF4uOGGGxg5ciS5ubmYzWbS0tKYOXMmS5cuZerUqZw4cYKxY8dSUlLCLbfcAsBHH33E0aNHUalUPPHEExQWFrJ//37eeecdJk+ezKlTp3jttddYv349Z8+exWq1MnfuXDQaDatXryY+Ph6r1drOpxUrVqDRaKipqeHll19m7dq1lJeXU1lZydNPP83WrVs5d+4cNpuNBx54gB07dnD27FkyMzMRRbFdP+F4rdgnRHjDDTfQ3NzMSy+9hMPhYNasWZw+fZrx48ejUqnYu3cvH3/8sVcsNTU1PPnkk177l156CaWy7c4lNzeX7OxsYmJiOHHiBLm5ud5ZhEuWLOHGG28kLS2N3/72tzz22GP87Gc/w+Px8N577zF+/Hhmz57NPffcw7PPPktxcbG3j7Fjx/LQQw/x1FNPUVVVxfr165k6dSpKpZLdu3dz+vRpsrOzMZlMVyyGWVJSwvXXX8+MGTNwOp0cOXKEN954wzuZ7PDhw6xatYr6+nqys7NJT0/nzjvvZPLkydx3333t+pFFGCI2b97M8OHDWbZsGW63m4ULF7Jy5UpWrlyJ3W4nKyuLrKws5s+fD7SNhDk5OR225XK5vHWfi4uLEQTB+1mhUCCKorfylUajQafT4XA4EAQBAIPBAIBWq/WOwABGo9FrY7VasVgsPP7445SXl1NUVMSZM2cQRRGFQtHOzuVy8cgjj+B0OsnNzWXRokX8EPptbm5u569SqfTui42NRRTFK/oJB31ChFlZWbz44ototVpcLhd33HEHFouFRYsWsXTpUgwGAw6Hg+nTpwNtbw8ef/xxr/3cuXOZNKltFbBFixZ5pyxMmTKFxYsXs2LFCuLj48nMzPRZ5PDTTz+lvLwci8XS6aw2k8nElClTeOqpp6ivr2fp0qUsWrSIF154AYvFgtPp9B6r0WhYv349Op0Os9nMiBEjGD16NCtWrKCmpoZly5Yxfvx4nn/+eVpbW1myZAnbt28H2goyXt5POJDfmMiEHTlEIxN2ZBHKhB1ZhDJhRxahTNiRRSgTdmQRyoQdWYQyYef/A7+KYqIFpYKhAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 144x144 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "func_cutoff = -.748\n",
    "lof_cutoff = -1.328\n",
    "\n",
    "fig,ax = plt.subplots(figsize=(2,2))\n",
    "sns.scatterplot(x='min_sge_score',y='AvgHAP1_MELJUSO;LFC_pdna;z-score',data=filtered,\n",
    "                hue='Mutation bin',palette=color_dict,s=12,linewidth=0,legend=False,\n",
    "                zorder=3)\n",
    "ax.axvline(x=func_cutoff,color='black',linestyle='dashed',linewidth=0.5,zorder=2)\n",
    "ax.axvline(x=lof_cutoff,color='black',linestyle='dashed',linewidth=0.5,zorder=2)\n",
    "ax.axhline(y=(-2),color='black',linestyle='dashed',linewidth=0.5,zorder=2)\n",
    "ax.axhline(y=(-4),color='black',linestyle='dashed',linewidth=0.5,zorder=2)\n",
    "ax.set_xlabel('SGE combined score',fontsize=8)\n",
    "plt.xticks(fontsize=8)\n",
    "plt.yticks([-10,-5,0],fontsize=8)   \n",
    "ax.set_ylabel('base editing, z-score',fontsize=8)\n",
    "for axis in ['top','bottom','left','right']:\n",
    "    ax.spines[axis].set_linewidth(0.5)\n",
    "    ax.tick_params(width=0.5,length=4)\n",
    "\n",
    "print 'Pearson r = ' + str(np.round(pearsonr(x=filtered['AvgHAP1_MELJUSO;LFC_pdna;z-score'],\n",
    "                                             y=filtered['min_sge_score'])[0],2))\n",
    "#fig.savefig('../Figures/Fig2_BRCA/Panels/SGE_comparison_v2.pdf',transparent=True,bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also determine the number of sgRNAs scoring in the different quadrants. Note that this doesn't add up to 51 because some sgRNAs score as \"intermediate\" hits in each dataset and therefore don't fall into one of the quadrants below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sgRNAs scoring in both datasets = 5\n",
      "Number of sgRNAs scoring in neither dataset = 24\n",
      "Number of sgRNAs scoring in base editor only = 2\n",
      "Number of sgRNAs scoring in SGE only = 8\n"
     ]
    }
   ],
   "source": [
    "print 'Number of sgRNAs scoring in both datasets = ' + str(len(filtered[(filtered['AvgHAP1_MELJUSO;LFC_pdna;z-score'] < -4) & (filtered['min_sge_score'] < lof_cutoff)]))\n",
    "print 'Number of sgRNAs scoring in neither dataset = ' + str(len(filtered[(filtered['AvgHAP1_MELJUSO;LFC_pdna;z-score'] > -2) & (filtered['min_sge_score'] > func_cutoff)]))\n",
    "print 'Number of sgRNAs scoring in base editor only = ' + str(len(filtered[(filtered['AvgHAP1_MELJUSO;LFC_pdna;z-score'] < -4) & (filtered['min_sge_score'] > func_cutoff)]))\n",
    "print 'Number of sgRNAs scoring in SGE only = ' + str(len(filtered[(filtered['AvgHAP1_MELJUSO;LFC_pdna;z-score'] > -2) & (filtered['min_sge_score'] < lof_cutoff)]))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
